go-controller/pkg/node/gateway.go

```diff
@@ -0,0 +1,98 @@
+package node
+
+import (
+	kapi "k8s.io/api/core/v1"
+)
+
+type ServiceEventHandler interface {
+	AddService(*kapi.Service)
+	DeleteService(*kapi.Service)
+	UpdateService(old, new *kapi.Service)
+	SyncServices([]interface{})
+}
+
+type EndpointsEventHandler interface {
+	AddEndpoints(*kapi.Endpoints)
+	DeleteEndpoints(*kapi.Endpoints)
+	UpdateEndpoints(old, new *kapi.Endpoints)
+}
+
+type ServiceAndEndpointsEventHandler interface {
+	ServiceEventHandler
+	EndpointsEventHandler
+}
+
+type Gateway interface {
+	ServiceAndEndpointsEventHandler
+	PostWait() postWaitFunc
+}
+
+type GatewayImpl interface {
+	ServiceEventHandler
+	PostWait() postWaitFunc
+}
+
+// gateway encapsulates state and functions that are shared between
+// local and shared gateway implementations
+
+type gateway struct {
+	gatewayImpl GatewayImpl
+	lbhc        ServiceAndEndpointsEventHandler
+	pcw         ServiceEventHandler
```

Comment:
Nothing ever treats `lbhc` as a "load balancer health checker" or "pcw" as a "port claim watcher". It seems like `gateway` is mostly just an array of `ServiceEventHandler` and an array of `EndpointEventHandler`...

Reply:
It is. The only reason I have these listed out explicitly vs. putting them in an array is so we can be explicit about the order that events are dispatched to each section.